File: design_pattern/kotlin/Builder.kt
package design_pattern.kotlin

data class Car(val brand: String, val model: String, val color: String, val engine: String, val transmission: String)

interface CarBuilder {
    fun setBrand(brand: String)

    fun setModel(model: String)

    fun setColor(color: String)

    fun setEngine(engine: String)

    fun setTransmission(transmission: String)

    fun build(): Car
}

class CarBuilderImpl : CarBuilder {
    private var brand: String = ""

    private var model: String = ""

    private var color: String = ""

    private var engine: String = ""

    private var transmission: String = ""

    override fun setBrand(brand: String) {
        this.brand = brand
    }

    override fun setModel(model: String) {
        this.model = model
    }

    override fun setColor(color: String) {
        this.color = color
    }

    override fun setEngine(engine: String) {
        this.engine = engine
    }

    override fun setTransmission(transmission: String) {
        this.transmission = transmission
    }

    override fun build(): Car {
        return Car(brand, model, color, engine, transmission)
    }
}

class CarDirector(private val builder: CarBuilder) {
    fun constructSportCar() {
        builder.setBrand("Ferrari")
        builder.setModel("488 GTB")
        builder.setColor("Red")
        builder.setEngine("V8")
        builder.setTransmission("7-Speed dual clutch")
    }

    fun constructSUV() {
        builder.setBrand("BMW")
        builder.setModel("X5")
        builder.setColor("Black")
        builder.setEngine("V6")
        builder.setTransmission("8-Speed automatic")
    }
}

fun main() {
    val builder = CarBuilderImpl()
    val director = CarDirector(builder)

    director.constructSportCar()
    val sportCar = builder.build()
    println(sportCar)

    director.constructSUV()
    val suvCar = builder.build()
    println(suvCar)
}

File: design_pattern/creational_patterns/dart/builder.dart
class Car {
  const Car(this.brand, this.model, this.engine, this.color, this.transmission);

  final String brand;
  final String model;
  final String engine;
  final String color;
  final String transmission;
}

abstract class CarBuilder {
  void setBrand(String brand);

  void setModel(String model);

  void setEngine(String engine);

  void setColor(String color);

  void setTransmission(String transmission);

  Car buildCar();
}

class CarBuilderImpl implements CarBuilder {
  String brand = '';

  String model = '';

  String engine = '';

  String color = '';

  String transmission = '';

  @override
  void setBrand(String brand) {
    this.brand = brand;
  }

  @override
  void setColor(String color) {
    this.color = color;
  }

  @override
  void setEngine(String engine) {
    this.engine = engine;
  }

  @override
  void setModel(String model) {
    this.model = model;
  }

  @override
  void setTransmission(String transmission) {
    this.transmission = transmission;
  }

  @override
  Car buildCar() {
    return Car(brand, model, engine, color, transmission);
  }
}

class CarDirector {
  CarDirector(this.builder);

  final CarBuilder builder;

  void constructSportCar() {
    builder
      ..setBrand('Ford')
      ..setEngine('V8')
      ..setColor('Black')
      ..setModel('S4')
      ..setTransmission('Transmission C7');
  }

  void constructSUV() {
    builder
      ..setBrand('Hyundai')
      ..setEngine('V6')
      ..setColor('Red')
      ..setModel('S2')
      ..setTransmission('Transmission C1');
  }
}

void main() {
  final builder = CarBuilderImpl();
  final director = CarDirector(builder);
  director.constructSportCar();
  final sportCar = builder.buildCar();
  print(sportCar);

  director.constructSUV();
  final suvCar = builder.buildCar();
  print(suvCar);
}
